controllers/cruisecontroltask_controller.go

```diff
@@ -70,121 +82,223 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
 			if task == nil {
 				continue
 			}
-			task.FromResult(result)
+			task.SetCruiseControlOperationRef(cruiseControlOpRef)
+			task.SetStateScheduled()
 		}
 	}
 
 	if err = r.UpdateStatus(ctx, instance, tasksAndStates); err != nil {
 		log.Error(err, "failed to update Kafka Cluster status")
 	}
-	return requeueAfter(DefaultRequeueAfterTimeInSec)
+
+	return reconciled()
+}
+
+func checkBrokersAvailability(scaler scale.CruiseControlScaler, brokerIDs []string) ([]string, error) {
+	states := []scale.KafkaBrokerState{scale.KafkaBrokerAlive, scale.KafkaBrokerNew}
+	// This can result NullPointerException when the capacity calculation is missing for a broker in the cruisecontrol configmap
+	availableBrokers, err := scaler.BrokersWithState(states...)
+	if err != nil {
+		if strings.Contains(err.Error(), nullPointerExceptionErrString) {
+			return nil, errors.WrapIff(err, "broker storage capacity calculations for Cruise Control has not been finished yet")
+		}
+		return nil, errors.WrapIff(err, "failed to retrieve list of available brokers from Cruise Control")
+	}
+
+	availableBrokersMap := scale.StringSliceToMap(availableBrokers)
+	unavailableBrokerIDs := make([]string, 0, len(brokerIDs))
+	for _, id := range brokerIDs {
+		if _, ok := availableBrokersMap[id]; !ok {
+			unavailableBrokerIDs = append(unavailableBrokerIDs, id)
+		}
+	}
+
+	return unavailableBrokerIDs, nil
 }
 
-// UpdateStatus updates the Status of the provided kafkav1beta1.KafkaCluster instance with the status of the tasks
+func (r *CruiseControlTaskReconciler) addBrokers(ctx context.Context, kafkaCluster *banzaiv1beta1.KafkaCluster, ttlSecondsAfterFinished *int, bokerIDs []string) (corev1.LocalObjectReference, error) {
+	return r.createCCOperation(ctx, kafkaCluster, banzaiv1alpha1.ErrorPolicyRetry, ttlSecondsAfterFinished, banzaiv1alpha1.OperationAddBroker, bokerIDs)
+}
+
+func (r *CruiseControlTaskReconciler) removeBroker(ctx context.Context, kafkaCluster *banzaiv1beta1.KafkaCluster, ttlSecondsAfterFinished *int, brokerID string) (corev1.LocalObjectReference, error) {
+	return r.createCCOperation(ctx, kafkaCluster, banzaiv1alpha1.ErrorPolicyRetry, ttlSecondsAfterFinished, banzaiv1alpha1.OperationRemoveBroker, []string{brokerID})
+}
+
+func (r *CruiseControlTaskReconciler) rebalanceDisks(ctx context.Context, kafkaCluster *banzaiv1beta1.KafkaCluster, ttlSecondsAfterFinished *int, bokerIDs []string) (corev1.LocalObjectReference, error) {
+	return r.createCCOperation(ctx, kafkaCluster, banzaiv1alpha1.ErrorPolicyRetry, ttlSecondsAfterFinished, banzaiv1alpha1.OperationRebalance, bokerIDs)
+}
+
+func (r *CruiseControlTaskReconciler) createCCOperation(ctx context.Context, kafkaCluster *banzaiv1beta1.KafkaCluster, errorPolicy banzaiv1alpha1.ErrorPolicyType, ttlSecondsAfterFinished *int,
+	operationType banzaiv1alpha1.CruiseControlTaskOperation, bokerIDs []string) (corev1.LocalObjectReference, error) {
```

Comment:
**Recommended**: I think in such long function name cases we **SHOULD** use the convention of splitting one parameter per line for readability/maintainability purposes. We might lose the omitted type possibility, but it makes function signatures more readable.

```suggestion
func (r *CruiseControlTaskReconciler) createCCOperation(
	ctx context.Context, 
	kafkaCluster *banzaiv1beta1.KafkaCluster, 
	errorPolicy banzaiv1alpha1.ErrorPolicyType,
	ttlSecondsAfterFinished *int,
	operationType banzaiv1alpha1.CruiseControlTaskOperation, 
	bokerIDs []string,
) (corev1.LocalObjectReference, error) {
```